app/models/user.rb

```diff
@@ -1,22 +1,43 @@
 class User < ApplicationRecord
-  before_save :downcase_email
+  attr_accessor :remember_token
 
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
 
   validates :name,  presence: true, length: { maximum: 50 }
   validates :email, presence: true, length: { maximum: 255 },
     format: { with: VALID_EMAIL_REGEX },
     uniqueness: { case_sensitive: false }
-  validates :password, presence: true, length: { minimum: 6 }
+  validates :password, presence: true, length: { minimum: 6 }, allow_nil: true
 
   has_secure_password
 
+  before_save :downcase_email
+
   class << self
     def digest string
       cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
         BCrypt::Engine.cost
       BCrypt::Password.create string, cost: cost
     end
+
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+
+  end
+
+  def remember
+    self.remember_token = self.class.new_token
+    update_attributes remember_digest: User.digest(remember_token)
+  end
+
+  def authenticated? remember_token
+    return false unless remember_digest
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attributes remember_digest: nil
```

Comment:
em phân biệt giúp anh mấy method này nhé:
```ruby
update_attribute
update_attributes
update
update_columns
```
Còn kiểu nào nữa ko nhỉ?

Reply:
ActiveRecord::Persistence instance methods
**update_attribute**(2 tham số là name và value) : 
  + cập nhật 1 trường của bản ghi,
  + trả về giá trị true/false
  + bỏ qua bước validation
  + có chạy callback (before_validation, after_validation, before_save, around_save, after_save, before_update, around_update, after_update)
  + cập nhật trường updated_at/updated_on (nếu có)
 **update_attributes** (alias for update) (1 tham số là 1 hash)
  + cập nhật nhiều trường của bản ghi
  + trả về giá trị true/false
  + raise exception nếu 1 trong các trường đó được mark read only
  + chạy callback và validation
  + cập nhật trường updated_at/updated_on (nếu có)
**update_columns** (1 tham số truyền vào là 1 hash)
  + cập nhật nhiều trường của bản ghi
  + bỏ qua bước validation
  + ko chạy callback
  + raise exception nếu 1 trong các trường đó được mark read only

còn update_attributes! (alias for update!) , update_column(name, value)
Và ActiveRecord::Persistence::ClassMethods class method update( 2 tham số là id và attributes: (1 hash)